docs/proposals/01-firmware-update.md

```diff
@@ -0,0 +1,381 @@
+  - `ScanResponse` MUST contain the status of the request with error code if any.
+- Discover(DiscoverRequest) DiscoverResponse;
+  - `DiscoverRequest` MUST contain the reference to concrete `DiscoveredFirmware` object.
+  - `DiscoverResponse` MUST contain the status of the request with error code if any.
+- UpdateServer(UpdateServerRequest) UpdateServerResponse;
+  - `UpdateServerRequest` MUST contain the reference to concrete `Server` object and the list of installed firmware-versions.
+  This endpoint MUST be used by update or scan jobs after a task if finished to send results and invoke the object's update.
+  - `UpdateServerResponse` MUST contain the status of the request with error code if any.
+- UpdateDiscoveredFirmware(UpdateDiscoveredFirmwareRequest) UpdateDiscoveredFirmwareResponse;
+  - `UpdateDiscoveredFirmwareRequest` MUST contain the reference to concrete `DiscoveredFirmware` object and the list of discovered firmware-versions.
+  This endpoint MUST be used by discovery jobs after a task is finished to send results and invoke the object's update.
+  - `UpdateDiscoveredFirmwareResponse` MUST contain the status of the request with error code if any.
+
+Depending on the type of the request, it SHOULD be forwarded to the corresponding scheduler's queue.
+
+#### Scheduler
+
+Scheduler is a component of the update service that is responsible for scheduling jobs:
+
+- it MUST NOT allow running several jobs on the same target server simultaneously;
+- it MAY discard incoming update or scan requests if the same jobs targeting the same server are already scheduled;
+- it MAY discard incoming discovery requests if the job targeting the same manufacturer-model pair is already scheduled;
+- it MUST have a mechanism to limit the number of parallel jobs;
+- it MUST have a mechanism to limit the job queue length;
+
+Scheduler SHOULD have an embedded job runner component corresponding to the update strategy defined on the update service application start.
+
+#### Job runner
+
+Job runner is a component that acts as a pool to spawn worker for a concrete job, considering the chosen update strategy and target server's manufacturer and model pair.
+
+- it MUST have a mechanism to store the metadata of spawned jobs;
+- it MUST have a mechanism to interrupt a running job;
+- it MAY have a mechanism to request the state of a long-running task;
+
+#### Update/scan/discovery job
+
+Standalone application which is a concrete implementation of firmware update/scan/discovery for a particular manufacturer or manufacturer-model pair.
+Implementation depends on the chosen update strategy.
+The application MUST run an API server providing the following endpoints:
+
+- CancelTask(CancelTaskRequest) CancelTaskResponse;
+  - `CancelTaskRequest` MAY contain the timeout for graceful stop and a flag to force stop.
+  - `CancelTaskResponse` MUST contain the status of the request with error code if any.
+
```

Comment:
Could we use this to pause upgrades if we find out that firmware has issues and investigation is needed before proceeding with next servers OR issues in cluster are noticed and we want to halt the upgrade process.

Reply:
This is the goal for proposed servers grouping - first run updates on dedicated test group. If all is ok, then run updates on prod servers.